charmcraft/commands/build.py

```diff
@@ -138,24 +140,79 @@ def __init__(self, args, config):
         self.ignore_rules = self._load_juju_ignore()
         self.config = config
 
-    def run(self):
-        """Build the charm."""
+    def build_charm(self, bases_config: Optional[BasesConfiguration]) -> str:
+        """Build the charm.
+
+        :param bases_config: Bases configuration to use for build.
+
+        :returns: File name of charm.
+        """
         logger.debug("Building charm in '%s'", self.buildpath)
 
         if self.buildpath.exists():
             shutil.rmtree(str(self.buildpath))
         self.buildpath.mkdir()
 
-        create_manifest(self.buildpath, self.config.project.started_at)
+        create_manifest(self.buildpath, self.config.project.started_at, bases_config)
 
         linked_entrypoint = self.handle_generic_paths()
         self.handle_dispatcher(linked_entrypoint)
         self.handle_dependencies()
-        zipname = self.handle_package()
+        zipname = self.handle_package(bases_config)
 
         logger.info("Created '%s'.", zipname)
         return zipname
 
+    def run(self) -> List[str]:
+        """Run build process.
+
+        In managed-mode (and eventually destructive-mode), build for each bases
+        configuration which has a matching build-on to the host we are executing
+        on.  Warn for each bases configuration that is unbuildable.  Error if
+        unable to produce any builds for any bases configuration.
+
+        Otherwise, build the charm in the legacy mode until we have complete
+        provider support and can drop it.
+
+        :returns: List of charm files created.
+        """
+        charms: List[str] = []
+
+        if is_charmcraft_running_in_managed_mode():
+            if self.config.bases is None:
+                # XXX: 2021-06-16 Patterson temporary until we have the default base
+                raise CommandError("Bases are currently required in managed-mode.")
+
+            for i, bases_config in enumerate(self.config.bases):
+                for j, build_on in enumerate(bases_config.build_on):
+                    matches, reason = check_if_base_matches_host(build_on)
+                    if matches:
+                        logger.debug(
```

Comment:
This and other log lines: do NOT compose the string yourself, but pass the template string as first parameter and the arguments as next parameters...

This is for performance and robustness.

For performance because if the logger decides that the message will not be sent, the string is not built at all. 

For robustness because in the event of introducing a bug in the line, the logging infrastructure will not crash, but report the problem. For example, in this case: `logger.info("Foo %s: %d", self.quantity)`.

Reply:
Indeed, pylint usually reminds me of that ;) will fix

Reply:
Done!